packages/yoastseo/src/parse/structure/Sentence.js

```diff
@@ -25,6 +30,60 @@ class Sentence {
 		 */
 		this.sourceCodeRange = {};
 	}
+
+	/**
+	 * Retrieves the first non-space token in the sentence.
+	 * @returns {Token|undefined} The first non-space token in the sentence.
+	 */
+	getFirstToken() {
+		return this.tokens.find( ( { text } ) => text !== " " );
+	}
+
+	/**
+	 * Retrieves the last non-space token in the sentence.
+	 * @returns {Token|undefined} The last non-space token in the sentence.
+	 */
+	getLastToken() {
+		return this.tokens.findLast( ( { text } ) => text !== " " );
+	}
+
+	/**
+	 * Sets the parent attributes for this sentence.
+	 * @param {Node} parentNode The parent node.
+	 * @param {Node} tree The full tree.
+	 */
+	setParentAttributes( parentNode, tree ) {
+		const node = parentNode;
+
+		// For implicit paragraphs, base the details on the parent of this node.
+		if ( parentNode.isImplicit ) {
+			parentNode = parentNode.getParentNode( tree );
+		}
+
+		/**
+		 * The start offset of the parent node.
+		 * The parent node's start offset is the start offset of the parent node if it doesn't have a `startTag` property.
+		 * @type {number}
+		 */
+		this.parentStartOffset = parentNode.getStartOffset();
+		/**
+		 * The block client ID of the parent node.
+		 * @type {string}
+		 */
+		this.parentClientId = parentNode.clientId || "";
+		/**
+		 * The attribute ID of the parent node, if available, otherwise an empty string.
+		 * Only used for position-based highlighting in sub-blocks of Yoast blocks.
+		 * @type {string}
+		 */
+		this.parentAttributeId = node.attributeId || "";
+		/**
+		 * Whether the parent node is the first section of Yoast sub-blocks.
+		 * Only used for position-based highlighting.
+		 * @type {boolean}
+		 */
+		this.isParentFirstSectionOfBlock = node.isFirstSection || false;
```

Comment:
After building the types, it's complaining that `clientId`, `isImplicit`, `attributeId`, and `isFirstSection` do not exist in Node. It doesn't need to be fixed in this PR and certainly can be a technical debt issue, but this generally makes me think that we should have a robust definition of a Class when we create one. For Node for example we also added those attributes when we parse the Gutenberg blocks. So either we initialize those attributes in Node's constructor or we create a typedef for Node along with its properties

Reply:
Yeah, let's not address this right now. I just tried it but it would break a lot of unit tests...